### PR TITLE
[Audio] Add webdataset in paddlespeech.audio

### PR DESCRIPTION
**Pull request by Jackwaterveg**
### PR types
New features
Add webdataset in paddlespeech.audio

### PR changes
Others

### Describe
1. 在 paddlespeech.audio 中加入 webdataset
2. 在 paddlespeech.audio 当中加入了 paddlespeech.s2t.transform

### CODE REVIEW

**Review by zh794390558 — COMMENTED** [review]
用streamdata目录吧。webdataset指定下版本号。

File: examples/wenetspeech/asr1/conf/conformer.yaml [hunk 1]
```diff
@@ -43,40 +42,54 @@ model_conf:
...
-batch_frames_inout: 0
-num_workers: 0
-subsampling_factor: 1
+batch_size: 32
+minlen_in: 10
+maxlen_in: 1200  # if input length(number of frames) > maxlen-in, data is automatically removed
+minlen_out: 0
+maxlen_out: 150  # if output length(number of tokens) > maxlen-out, data is automatically removed
+resample_rate: 16000
+shuffle_size: 1500
```
Comment (zh794390558) [inline examples/wenetspeech/asr1/conf/conformer.yaml]: shuffle_size
sort_size
加上注释说明

File: examples/wenetspeech/asr1/conf/conformer.yaml [hunk 2]
```diff
@@ -43,40 +42,54 @@ model_conf:
...
+sort_size: 1000
+num_workers: 8
+prefetch_factor: 10
+dist_sampler: True
 num_encs: 1
+augment_conf:
```
Comment (zh794390558) [inline examples/wenetspeech/asr1/conf/conformer.yaml]: augment_conf
这些参数是否能从preprocess.conf中获取？

File: examples/wenetspeech/asr1/conf/conformer.yaml [hunk 3]
```diff
@@ -43,40 +42,54 @@ model_conf:
...
-batch_frames_out: 0
-batch_frames_inout: 0
-num_workers: 0
-subsampling_factor: 1
+batch_size: 32
+minlen_in: 10
+maxlen_in: 1200  # if input length(number of frames) > maxlen-in, data is automatically removed
+minlen_out: 0
+maxlen_out: 150  # if output length(number of tokens) > maxlen-out, data is automatically removed
+resample_rate: 16000
```
Comment (zh794390558) [inline examples/wenetspeech/asr1/conf/conformer.yaml:68]: 这些参数是否能从preprocess.conf中获取？

File: examples/wenetspeech/asr1/local/data.sh [hunk 4]
```diff
@@ -44,86 +46,57 @@ if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
...
+  echo "Making shards, please wait..."
+  RED='\033[0;31m'
+  NOCOLOR='\033[0m'
+  echo -e "It requires ${RED}1.2T ${NOCOLOR}space for $shards_dir, please make sure you have enough space"
+  echo -e "It takes about ${RED}12 ${NOCOLOR}hours with 32 threads"
+  for x in $dev_set $test_sets ${train_set}; do
+    dst=$shards_dir/$x
+    mkdir -p $dst
+    utils/make_filted_shard_list.py --resample 16000 --num_utts_per_shard 1000 \
+      --do_filter --num_node 1 --num_gpus_per_node 8 \
```
Comment (zh794390558) [inline examples/wenetspeech/asr1/local/data.sh]: --num_node 1 --num_gpus_per_node 8 参数化，放在开头。

File: paddlespeech/audio/streamdata/__init__.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
+# Copyright (c) 2017-2019 NVIDIA CORPORATION. All rights reserved.
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/__init__.py:1]: 确认我们有改动的，可以加上自己的short copyright

File: paddlespeech/audio/streamdata/autodecode.py [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+# Copyright (c) 2017-2021 NVIDIA CORPORATION. All rights reserved.
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+# This file is part of the WebDataset library.
+# See the LICENSE file for licensing terms (BSD-style).
+# Modified from https://github.com/webdataset/webdataset
+#
+
+"""Automatically decode webdataset samples."""
+
+import io, json, os, pickle, re, tempfile
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/autodecode.py:11]: 需要format下code

File: paddlespeech/audio/streamdata/autodecode.py [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+def cbor_loads(data):
+    import cbor
+
+    return cbor.loads(data)
+
+
+decoders = {
+    "txt": lambda data: data.decode("utf-8"),
+    "text": lambda data: data.decode("utf-8"),
+    "transcript": lambda data: data.decode("utf-8"),
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/autodecode.py:66]: 增加 translation

File: paddlespeech/audio/streamdata/autodecode.py [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+    "cls": lambda data: int(data),
+    "cls2": lambda data: int(data),
+    "index": lambda data: int(data),
+    "inx": lambda data: int(data),
+    "id": lambda data: int(data),
+    "json": lambda data: json.loads(data),
+    "jsn": lambda data: json.loads(data),
+    "pyd": lambda data: pickle.loads(data),
+    "pickle": lambda data: pickle.loads(data),
+    "pdparams": lambda data: paddle_loads(data),
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/autodecode.py:76]: pdparams -> pth

File: paddlespeech/audio/streamdata/autodecode.py [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+    """Decode audio using the paddleaudio library.
+
+    :param key: file name extension
+    :param data: data to be decoded
+    """
+    extension = re.sub(r".*[.]", "", key)
+    if extension not in ["flac", "mp3", "sox", "wav", "m4a", "ogg", "wma"]:
+        return None
+
+    import paddleaudio
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/autodecode.py]: 用paddlespeech.audio。paddleaudio废弃了

File: paddlespeech/audio/streamdata/cache.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+                with open(dest, "rb") as f:
+                    data = f.read(200)
+                os.remove(dest)
+                raise ValueError(
+                    "%s (%s) is not a tar archive, but a %s, contains %s"
+                    % (dest, url, ftype, repr(data))
+                )
+            try:
+                stream = open(dest, "rb")
+                sample.update(stream=stream)
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/cache.py:151]: 注明输出的格式是： dict(url=url， stream=stream)

File: paddlespeech/audio/streamdata/tariterators.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    for source in data:
+        url = source["url"]
+        try:
+            assert isinstance(source, dict)
+            assert "stream" in source
+            for sample in tar_file_and_group_iterator(source["stream"]):
+                assert (
+                    isinstance(sample, dict) and "wav" in sample and "txt" in sample and "fname" in sample
+                )
+                sample["__url__"] = url
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/tariterators.py:232]: sample的输入输出格式在docstring里注明，便于理解代码。

File: paddlespeech/audio/streamdata/compat.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+        if size < 1:
+            return self
+        else:
+            return self.compose(filters.shuffle(size, **kw))
+
+    def map(self, f, handler=reraise_exception):
+        return self.compose(filters.map(f, handler=handler))
+
+    def decode(self, *args, pre=None, post=None, only=None, partial=False, handler=reraise_exception):
+        handlers = [autodecode.ImageHandler(x) if isinstance(x, str) else x for x in args]
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/compat.py:47]: args是handlers的list，可以加个选项。例如audio，true的时候创建AudioHandler。

File: paddlespeech/audio/streamdata/compat.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    def xdecode(self, *args, **kw):
+        return self.compose(filters.xdecode(*args, **kw))
+
+    def data_filter(self, *args, **kw):
+        return self.compose(filters.data_filter(*args, **kw))
+    
+    def tokenize(self, *args, **kw):
+        return self.compose(filters.tokenize(*args, **kw))
+
+    def resample(self, *args, **kw):
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/compat.py:87]: 语音相关的可以放在一起。应该可以复用decode中的Decoder的逻辑。

File: paddlespeech/audio/streamdata/extradatasets.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        self.sample = sample
+        self.length = length
+
+    def __iter__(self):
+        """Return an iterator over this mock dataset."""
+        for i in range(self.length):
+            yield self.sample
+
+
+class repeatedly(IterableDataset, PipelineStage):
```
Comment (zh794390558) [inline paddlespeech/audio/streamdata/extradatasets.py:48]: 这些接口是否能解决dist.join的问题？

File: paddlespeech/audio/utils/dynamic_import.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+    :param str import_path: syntax 'module_name:class_name'
+        e.g., 'paddlespeech.s2t.models.u2:U2Model'
+    :param dict alias: shortcut for registered class
+    :return: imported class
+    """
+    if import_path not in alias and ":" not in import_path:
+        raise ValueError(
+            "import_path should be one of {} or "
+            'include ":", e.g. "paddlespeech.s2t.models.u2:U2Model" : '
```
Comment (zh794390558) [inline paddlespeech/audio/utils/dynamic_import.py:31]: paddlespeech.s2t.models.u2:U2Model

File: paddlespeech/audio/utils/dynamic_import.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+import importlib
+
+__all__ = ["dynamic_import"]
+
+
+def dynamic_import(import_path, alias=dict()):
+    """dynamic import module and class
+
+    :param str import_path: syntax 'module_name:class_name'
+        e.g., 'paddlespeech.s2t.models.u2:U2Model'
```
Comment (zh794390558) [inline paddlespeech/audio/utils/dynamic_import.py:24]: paddlespeech.s2t.models.u2:U2Model

File: paddlespeech/audio/utils/tensor_utils.py [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unility functions for Transformer."""
+from typing import List
+from typing import Tuple
+
+import paddle
+
+from .log import Logger
```
Comment (zh794390558) [inline paddlespeech/audio/utils/tensor_utils.py:20]: 如果s2t里用到这些函数了，就s2t依赖audio里的吧。

File: paddlespeech/s2t/exps/u2/model.py [hunk 18]
```diff
@@ -132,7 +134,7 @@ def valid(self):
                 msg = f"Valid: Rank: {dist.get_rank()}, "
                 msg += "epoch: {}, ".format(self.epoch)
                 msg += "step: {}, ".format(self.iteration)
-                msg += "batch: {}/{}, ".format(i + 1, len(self.valid_loader))
+                #msg += "batch: {}/{}, ".format(i + 1, len(self.valid_loader))
```
Comment (zh794390558) [inline paddlespeech/s2t/exps/u2/model.py]: if not self.use_streamdata 判断下

File: paddlespeech/s2t/exps/u2/model.py [hunk 19]
```diff
@@ -218,92 +222,188 @@ def do_train(self):
...
-                batch_frames_out=0,
-                batch_frames_inout=0,
-                preprocess_conf=config.preprocess_config,
-                n_iter_processes=config.num_workers,
-                subsampling_factor=1,
-                num_encs=1,
-                dist_sampler=config.get('dist_sampler', False),
-                shortest_first=False)
+            if self.use_streamdata:
+                self.train_loader = StreamDataLoader(
```
Comment (zh794390558) [inline paddlespeech/s2t/exps/u2/model.py]: 写个工厂函数吧。

File: paddlespeech/s2t/io/dataloader.py [hunk 20]
```diff
@@ -28,6 +29,9 @@
 from paddlespeech.s2t.io.reader import LoadInputsAndTargets
 from paddlespeech.s2t.utils.log import Log
 
+import paddlespeech.audio.streamdata as streamdata
+from paddlespeech.s2t.frontend.featurizer.text_featurizer import TextFeaturizer
```
Comment (zh794390558) [inline paddlespeech/s2t/io/dataloader.py]: TextFeaturizer挪到audio.text中

File: paddlespeech/s2t/io/dataloader.py [hunk 21]
```diff
@@ -56,6 +60,98 @@ def batch_collate(x):
...
+        self.feat_dim = num_mel_bins 
+        self.vocab_size = text_featurizer.vocab_size 
+        
+        # The list of shard
+        shardlist = []
+        with open(manifest_file, "r") as f:
+            for line in f.readlines():
+                shardlist.append(line.strip())
+        
+        if self.dist_sampler:
```
Comment (zh794390558) [inline paddlespeech/s2t/io/dataloader.py:143]: base_dataset 可以合成一个

File: paddlespeech/s2t/io/dataloader.py [hunk 22]
```diff
@@ -56,6 +60,98 @@ def batch_collate(x):
...
+            base_dataset = streamdata.DataPipeline(
+                streamdata.SimpleShardList(shardlist),
+                streamdata.split_by_worker,
+                streamdata.tarfile_to_samples(streamdata.reraise_exception)
+            )
+
+        self.dataset = base_dataset.append_list(
+            streamdata.tokenize(symbol_table),
+            streamdata.data_filter(frame_shift=frame_shift, max_length=maxlen_in, min_length=minlen_in, token_max_length=maxlen_out, token_min_length=minlen_in),
+            streamdata.resample(resample_rate=resample_rate),
```
Comment (zh794390558) [inline paddlespeech/s2t/io/dataloader.py]: resample 变成 audio_resample

File: paddlespeech/s2t/io/dataloader.py [hunk 23]
```diff
@@ -56,6 +60,98 @@ def batch_collate(x):
...
+        else:
+            base_dataset = streamdata.DataPipeline(
+                streamdata.SimpleShardList(shardlist),
+                streamdata.split_by_worker,
+                streamdata.tarfile_to_samples(streamdata.reraise_exception)
+            )
+
+        self.dataset = base_dataset.append_list(
+            streamdata.tokenize(symbol_table),
+            streamdata.data_filter(frame_shift=frame_shift, max_length=maxlen_in, min_length=minlen_in, token_max_length=maxlen_out, token_min_length=minlen_in),
```
Comment (zh794390558) [inline paddlespeech/s2t/io/dataloader.py]: data_filter -> audio_data_filter

File: paddlespeech/s2t/io/dataloader.py (hunk 22 above)
Comment (zh794390558) [reply paddlespeech/s2t/io/dataloader.py]: 要不都带上audio_吧

File: examples/wenetspeech/asr1/local/data.sh (hunk 4 above)
Comment (zh794390558) [reply examples/wenetspeech/asr1/local/data.sh]: make_filted_shard_list.py 没有上传

File: paddlespeech/audio/utils/dynamic_import.py (hunk 15 above)
Comment (Jackwaterveg) [reply paddlespeech/audio/utils/dynamic_import.py:31]: paddlespeech.s2t.models.u2:U2Model

File: examples/wenetspeech/asr1/conf/conformer.yaml [hunk 24]
```diff
@@ -42,7 +42,7 @@ model_conf:
 ###########################################
 #                   Data                  #
 ###########################################
-train_manifest: data/train_l/data.list
+train_manifest: data/train_l/data_tiny2.list
```
Comment (zh794390558) [inline examples/wenetspeech/asr1/conf/conformer.yaml]: 修改回默认值。

**Review by zh794390558 — APPROVED** [review]
LGTM
